### PR TITLE
Added multi-criteria airdrop engine with activity-based scoring and seamless AirdropManager integration

### PR DESCRIPTION
**Pull request by Arowolokehinde**
## Added Multi-Criteria Airdrop Engine with Activity-Based Scoring

## Overview

This PR introduces a Multi-Criteria Airdrop Engine that extends the existing RootStock airdrop template with advanced
eligibility mechanisms. The implementation focuses on activity-based scoring while providing a flexible framework for future
integration with staking protocols and RNS domains.

###  Multi-Criteria Scoring System

- Activity-based eligibility: Transaction count and contract interaction tracking
- Weighted scoring algorithm: Configurable weights for different criteria types
- Flexible architecture: Ready for staking and RNS domain integration
- Real-time scoring: Dynamic eligibility calculation with event emission

###  Advanced Functionality

- Batch operations: Gas-efficient bulk transaction and interaction tracking
- Admin controls: Dynamic criteria updates with validation
- Score transparency: Detailed scoring events for off-chain analytics
- Integration ready: Seamless compatibility with existing AirdropManager

### New Files

- `contracts/MultiCriteriaAirdrop1155.sol` (340+ lines)  
  - Core multi-criteria airdrop contract  
  - Implements IAirdrop1155 interface for full compatibility  
  - Activity tracking with batch operations  
  - Configurable scoring weights and thresholds  

- `test/MultiCriteriaAirdrop.test.ts` (236 lines)  
  - Comprehensive test suite with 15 test cases  
  - Activity scoring validation  
  - AirdropManager integration tests  
  - Admin functionality and edge cases  

### Modified Files

- `test/AirdropManager.test.ts`  
  - Fixed constructor parameter compatibility  
  - Added missing AirdropType parameter  
  - Updated claim function calls with required parameters  

### Core Components

```solidity
struct EligibilityCriteria {
    // Staking criteria (ready for integration)
    address stakingContract;
    uint256 minimumStakeAmount;
    uint256 minimumStakeDuration;

    // Activity criteria (implemented)
    uint256 minimumTransactions;
    uint256 minimumContractInteractions;

    // RNS criteria (ready for integration)
    address rnsRegistry;
    bytes32[] requiredDomains;
    bool requireAnyDomain;

    // Scoring configuration
    uint256 stakingWeight;    // % weight (0-100)
    uint256 activityWeight;   // % weight (0-100)
    uint256 rnsWeight;        // % weight (0-100)
    uint256 minimumScore;     // Required total score
}
````

### Scoring Algorithm

* Transaction Score: Based on volume and frequency (0-50 points)
* Contract Interaction Score: Based on DeFi/ecosystem engagement (0-50 points)
* Total Score: `(activityScore * activityWeight) / 100`
* Eligibility: `totalScore >= minimumScore`

### Future-Ready Design

* Staking protocol integration interfaces
*  RNS domain ownership verification
*  Cross-chain compatibility hooks
*  Advanced analytics and reporting

##  Test Coverage

15 tests passing
├── Deployment and Configuration (2 tests)
├── Activity Score Calculation (2 tests)
├── AirdropManager Integration (3 tests)
└── Admin Functions (2 tests)

Coverage: Core functionality, edge cases, integration

### Use Cases

* Governance Token Distribution: Reward active ecosystem participants
* DeFi Protocol Incentives: Target users based on interaction patterns
* Community Building: Incentivize cross-protocol engagement
* Developer Adoption: Reward contract interaction and usage

### CODE REVIEW

File: contracts/MultiCriteriaAirdrop1155.sol [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+        eligibilityCriteria = criteria;
+        
+        // Ensure weights add up to 100
+        require(
+            criteria.stakingWeight + criteria.activityWeight + criteria.rnsWeight == 100,
+            "Weights must sum to 100"
+        );
+    }
+
+    function claim(address user, uint256 amount, bytes32[] calldata proof) public onlyOwner {
```
Comment (scguaquetam) [inline contracts/MultiCriteriaAirdrop1155.sol]: Hello @Arowolokehinde , on my test and compilation, I found this  ,could you take a look? looks like there are some unused parameters which we should not let there.

Comment (scguaquetam) [reply contracts/MultiCriteriaAirdrop1155.sol]: Image

File: contracts/MultiCriteriaAirdrop1155.sol [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+        userScores[_address] = 0;
+    }
+
+    function disallowAddresses(address[] memory addresses) external onlyOwner {
+        for (uint i = 0; i < addresses.length; i++) {
+            userScores[addresses[i]] = 0;
+        }
+    }
+
+    function setRoot(bytes32 _root) external onlyOwner {
```
Comment (scguaquetam) [inline contracts/MultiCriteriaAirdrop1155.sol]: Take a look at this one also

Image

File: contracts/MultiCriteriaAirdrop1155.sol [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+    function getAirdropAmountLeft() public view returns(uint256) {
+        return _airdropAmountLeft;
+    }
+
+    function getBalance() public view returns(uint256) {
+        return _tokenContract.balanceOf(address(this), _tokenId);
+    }
+
+    function onERC1155Received(address operator, address from, uint256 id, uint256 value, bytes memory data) external pure returns (bytes4) {
```
Comment (scguaquetam) [inline contracts/MultiCriteriaAirdrop1155.sol]: This one also

Image

**Comment by Arowolokehinde** [comment]
i have fixed the issue with the unused parameters.
Thanks

**Comment by Arowolokehinde** [comment]
I have made the necessary changes.
Thank you

File: package.json [hunk 4]
```diff
@@ -31,5 +31,6 @@
     "hardhat-preprocessor": "^0.1.5",
     "keccak256": "^1.0.6",
     "merkletreejs": "^0.4.0"
-  }
+  },
+  "packageManager": "yarn@1.22.22+sha512.a6b2f7906b721bba3d67d4aff083df04dad64c399707841b7acf00f6b133b7ac24255f2652fa22ae3534329dc6180534e98d17432037ff6fd140556e2bb3137e"
```
Comment (scguaquetam) [inline package.json]: Is this one needed? @Arowolokehinde

Comment (Arowolokehinde) [reply package.json]: Sincerely apologize for the late response. My laptop spoilt unexpectedly during the week. 

This was auto-generated when I ran yarn install on my machine. It's not needed so i already removed it .

Comment (scguaquetam) [reply package.json]: Cool, no issue, thanks.

**Comment by Arowolokehinde** [comment]
Awesome, thanks!

**Comment by Arowolokehinde** [comment]
Are you still reviewing the PR?

File: package.json (hunk 4 above)
Comment (Arowolokehinde) [reply package.json]: Would like to ask if there’s anything else I need to do to ensure my PR gets merged

**Comment by scguaquetam** [comment]
Hello, everything is okay on your end, we are checking this now with the people in charge.

**Review by ezequiel-rodriguez — APPROVED** [review]
LGTM

**Comment by ezequiel-rodriguez** [comment]
Hey @Arowolokehinde,
Thanks a lot for your contribution! 
Someone from the team will be reaching out to you soon so you can receive your reward.

Keep it up,
cheers!

**Comment by Arowolokehinde** [comment]
Awesome, thank you so much!
